### PR TITLE
Add spring-boot-starter-camunda-sdk module

### LINKED ISSUE #16876: Migrate spring boot starter and remove deprecated code
**Description**
- Migrate `spring-boot-starter-camunda-sdk`
- Remove Operate and Tasklist related code
- Remove unused code from `spring-boot-starter-camunda-sdk`

### PR DESCRIPTION
**Pull request by ana-vinogradova-camunda**
## Description

- Migrate spring-boot-starter-camunda-sdk to zeebe
- Remove Operate and Tasklist related code
- Remove unused code from spring-boot-starter-camunda-sdk
- Adjust the licenses to match our client licenses
- Refactor code to pass Java checks

## Related issues

closes #16876

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: spring-boot-starter-camunda-sdk/pom.xml [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>${version.slf4j-api}</version>
+    </dependency>
+
+    <dependency>
+      <groupId>io.camunda.spring</groupId>
+      <artifactId>spring-client-zeebe</artifactId>
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/pom.xml:55]: should be `zeebe-client-spring` now, this is the old module from the community project

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/event/ZeebeLifecycleEventProducer.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  public ZeebeLifecycleEventProducer(
+      final ZeebeClient client, final ApplicationEventPublisher publisher) {
+    this.client = client;
+    this.publisher = publisher;
+  }
+
+  @Override
+  public void start() {
+    publisher.publishEvent(
+        new ClientStartedEvent()); // keep old deprecated event for a bit before delting it
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/event/ZeebeLifecycleEventProducer.java:39]: ❌  deprecated event
```suggestion
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/event/ZeebeLifecycleEventProducer.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    publisher.publishEvent(new ZeebeClientCreatedEvent(this, client));
+
+    running = true;
+  }
+
+  @Override
+  public void stop() {
+    publisher.publishEvent(
+        new ClientStoppedEvent()); // keep old deprecated event for a bit before delting it
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/event/ZeebeLifecycleEventProducer.java:48]: ❌  deprecated event
```suggestion
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/actuator/MicrometerMetricsRecorder.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MicrometerMetricsRecorder implements MetricsRecorder {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/actuator/MicrometerMetricsRecorder.java:34]:
```suggestion
      LoggerFactory.getLogger(MicrometerMetricsRecorder.class);
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    prefix = "zeebe.client",
+    name = "enabled",
+    havingValue = "true",
+    matchIfMissing = true)
+@ConditionalOnMissingBean(SpringZeebeTestContext.class)
+@ImportAutoConfiguration({ExecutorServiceConfiguration.class, ZeebeActuatorConfiguration.class})
+@AutoConfigureBefore(ZeebeClientAllAutoConfiguration.class)
+public class ZeebeClientProdAutoConfiguration {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java:47]:
```suggestion
  private static final Logger LOG = LoggerFactory.getLogger(ZeebeClientProdAutoConfiguration.class);
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    builder.withJsonMapper(jsonMapper);
+    final List<ClientInterceptor> legacyInterceptors = configurationProperties.getInterceptors();
+    if (!legacyInterceptors.isEmpty()) {
+      builder.withInterceptors(legacyInterceptors.toArray(new ClientInterceptor[0]));
+    } else if (clientInterceptorList != null && !clientInterceptorList.isEmpty()) {
+      builder.withInterceptors(clientInterceptorList.toArray(new ClientInterceptor[0]));
+    }
+    return builder;
+  }*/
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java:98]: ❌ 
```suggestion
```

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/properties/CommonConfigurationProperties.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.NestedConfigurationProperty;
+
+@ConfigurationProperties(prefix = "common")
+public class CommonConfigurationProperties extends Client {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/properties/CommonConfigurationProperties.java:29]: ❌  unused
```suggestion
```

**Review by megglos — CHANGES_REQUESTED** [review]
I didn't manage to go through all yet, the major change needed is actually making use of the new `zeebe-client-spring` module you added before, this change still pulls in the module from the community project.

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java (hunk 6 above)
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientProdAutoConfiguration.java:98]: I've removed that before, not sure why it appeared again... Will fix, thank you

File: spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  private static ZeebeClientConfigurationProperties properties() {
+    final ZeebeClientConfigurationProperties properties =
+        new ZeebeClientConfigurationProperties(null);
+    properties.applyOverrides();
+    return properties;
+  }
+
+  @JobWorker
+  void sampleWorker(
+      @Variable final String var1, @VariablesAsType final ComplexProcessVariable var2) {}
```
[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java:58]: Needed for tests.

[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java:58]: Needed for tests.

File: spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+  @JobWorker
+  void activatedJobWorker(@Variable final String var1, final ActivatedJob activatedJob) {}
```
[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java:61]: Needed for tests.

[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/properties/PropertyBasedZeebeWorkerValueCustomizerTest.java:61]: Needed for tests.

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        + ", commonConfigurationProperties="
+        + commonConfigurationProperties
+        + ", authentication="
+        + authentication
+        + ", jsonMapper="
+        + jsonMapper
+        + ", interceptors="
+        + interceptors
+        + ", zeebeClientExecutorService="
+        + zeebeClientExecutorService
```
[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:267]: `ZeebeClientExecutorService` is a part of `zeebe-client-spring` that we are not touching in this PR

File: spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/config/ZeebeClientStarterAutoConfigurationCustomJsonMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  void testBuilder() {
+    // ZeebeClientBuilder builder = autoConfiguration.builder(jsonMapper, Collections.emptyList());
+    // assertThat(builder).isNotNull();
+
+    final ZeebeClient client = applicationContext.getBean(ZeebeClient.class);
+    final io.camunda.zeebe.client.api.JsonMapper clientJsonMapper =
+        AopTestUtils.getUltimateTargetObject(client.getConfiguration().getJsonMapper());
+    assertThat(clientJsonMapper).isSameAs(jsonMapper);
+    assertThat(clientJsonMapper).isSameAs(applicationContext.getBean("overridingJsonMapper"));
+    assertThat(client.getConfiguration().getGatewayAddress()).isEqualTo("localhost12345");
```
[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/test/java/io/camunda/zeebe/spring/client/config/ZeebeClientStarterAutoConfigurationCustomJsonMapperTest.java]: Fixed

File: spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+import jakarta.annotation.PostConstruct;
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+
+public class ZeebeClientConfiguration implements io.camunda.zeebe.client.ZeebeClientConfiguration {
```
[the remark this answers is not in the document]
Comment (ana-vinogradova-camunda) [reply spring-boot-starter-camunda-sdk/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java:44]: ZeebeClientConfiguration that is being implemented comes from zeebe-client-java package.
We might consider renaming the class in question at a later stage, I took notes of this issue.

**Review by megglos — APPROVED** [review]
🚀  Great you could strip this project further down.
Lots to cleanup still to do, but it's ok to follow-up after 8.5.

🔧  The Product enum still being present should be something we cleanup next to reduce the overhead by the unnecessary classes related to supporting multiple app auth tokens.

🔧  I can prepare the REST config change for our testing next Monday.

💭  some commit types are actually style and not feat as we discussed on the call, no need to rename them now but let's keep that in mind for future changes

File: spring-boot-starter-camunda-sdk/pom.xml [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+  <licenses>
+    <license>
+      <name>Apache License, Version 2.0</name>
+      <url>http://www.apache.org/licenses/LICENSE-2.0</url>
+    </license>
+  </licenses>
+
+  <properties>
+    <version.slf4j-api>2.0.12</version.slf4j-api>
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/pom.xml:36]: Will push that separately before merge
```suggestion
```

File: spring-boot-starter-camunda-sdk/pom.xml [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+    <license.header>com/mycila/maven/plugin/license/templates/APACHE-2.txt</license.header>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>${version.slf4j-api}</version>
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/pom.xml:44]:
```suggestion
```

File: spring-boot-starter-camunda-sdk/pom.xml [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+  </dependencies>
+
+  <build>
+    <resources>
+      <!-- we need to filter the resources, so placeholders in banner.txt get replaced -->
+      <resource>
+        <filtering>true</filtering>
+        <directory>src/main/resources</directory>
+      </resource>
+    </resources>
```
Comment (megglos) [inline spring-boot-starter-camunda-sdk/pom.xml:175]: ❌  this seems outdated, there is no banner.txt file anymore
```suggestion
```

**Pull request merged**